### ISSUE TITLE
WarningsAsErrors for CS8602 is not returning an error

### ISSUE THREAD

**Issue by J0rgeSerran0** [dotnet/roslyn#38371]
**Version Used**: 
Visual Studio 2019 16.2.3
.NET Core: 3.0.0-preview8-28405-07

**Steps to Reproduce**:

For CS8602 I am not getting the expected errors when I add this code in the WarningsAsErrors section.

1. Modify the csproj as:

&lt;Project Sdk="Microsoft.NET.Sdk"&gt;
  &lt;PropertyGroup&gt;
    &lt;OutputType&gt;Exe&lt;/OutputType&gt;
    &lt;TargetFramework&gt;netcoreapp3.0&lt;/TargetFramework&gt;
    **&lt;Nullable&gt;enable&lt;/Nullable&gt;
    &lt;WarningsAsErrors&gt;CS8602&lt;/WarningsAsErrors&gt;**
  &lt;/PropertyGroup&gt;
&lt;/Project&gt;

2. My sample application code is:

`private static void PosibleDereferenceNullableReferenceTypes()
{
     string data = null;
     Console.WriteLine(data.Length);
}`

**Expected Behavior**:
An error when we build the project.
**Error CS8602 Dereference of a possibly null reference.**

> Note: When I put the codes **CS8600, CS8603 or CS8604** into the **WarningsAsErrors** section, I receive the expected error messages in compilation time.

**Actual Behavior**:
Warning CS8602 Dereference of a possibly null reference.

**Comment by J0rgeSerran0** [comment]
## Additional notes

If I change my csproj as:

&lt;Project Sdk="Microsoft.NET.Sdk"&gt;
  &lt;PropertyGroup&gt;
    &lt;OutputType&gt;Exe&lt;/OutputType&gt;
    &lt;TargetFramework&gt;netcoreapp3.0&lt;/TargetFramework&gt;
    **&lt;TreatWarningsAsErrors&gt;true&lt;/TreatWarningsAsErrors&gt;**
  &lt;/PropertyGroup&gt;
&lt;/Project&gt;

And my application code is:

private static void PosibleDereferenceNullableReferenceTypes()
{
    #nullable enable
    string data = null;
    Console.WriteLine(data.Length);
    #nullable restore
}

**Here I get the error for CS8602 as expected**

**Comment by RikkiGibson** [comment]
For some reason, the command line build fails, but we don't get an error in the VS error list. Looking into whether there is maybe a bug in the VS workspace.

**Comment by RikkiGibson** [comment]
This may be an instance of dotnet/project-system#3680, but relating to the WarningsAsErrors property instead of the WarningsNotAsErrors property.

**Comment by RikkiGibson** [comment]
Basically, it seems like for some values of WarningsAsErrors, the project system sends us a command line with `... /warnAsError- ... /warnAsError+:CS0105 ...`. For other values it sends us a command line like `... /warnAsError+:CS8602 ... /warnAsError- ...`. The latter case is where the bug happens-- warnAsError- wins and we don't get the right behavior in the IDE.

**Comment by jcouv** [comment]
https://github.com/dotnet/roslyn/issues/52407 (issue with `WarningsAsErrors=nullable` setting) indeed looks to be caused by the same random ordering problem :-/

**Comment by RikkiGibson** [comment]
We should look into whether we can work around the issue by eliding `/warnAsError-`. That is, in the Csc task, can we make it so `<TreatWarningsAsErrors>false<TreatWarningsAsErrors>` results in **no** argument getting added to the command line, while `<TreatWarningsAsErrors>true</TreatWarningsAsErrors>` results in `/warnAsError+` getting added to the command line. /cc @jaredpar @dotnet/roslyn-compiler, and work around this problem without changing the semantics of builds.

**Comment by MatisseHack** [comment]
I'm running into this same issue in VS Code (OmniSharp) when trying to use `<WarningsAsErrors>nullable</WarningsAsErrors>`

**Comment by jcouv** [comment]
FWIW, I've tried this just now on VS 17.0 and I find that the issue was fixed. That matches my expectation, since the underlying project-system issue was fixed (either in 16.10 or 17.0, I'm not sure).

When I tried in VS Code however, there indeed still appears to be a problem (squiggles showing as warning instead of error):
![image](https://user-images.githubusercontent.com/12466233/144114124-d890cee6-c687-441c-8a03-eed022cebb81.png)

**Comment by RikkiGibson** [comment]
Hmm, I think the root cause was in internal VS code. So is it possible that OmniSharp has independently introduced a similar bug in its project system?

**Comment by RikkiGibson** [comment]
Maybe @JoeRobich or @333fred would know what corresponding place in OmniSharp could have this issue. Anywhere that compiler command line arguments get sliced and diced and reconstituted, etc.

**Comment by JoeRobich** [comment]
For OmniSharp, take a look in this file https://github.com/OmniSharp/omnisharp-roslyn/blob/c5f10ff18e4e0528ce54e71c21e62e4a56c32a8f/src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfoExtensions.cs#L61-L88

**Comment by MatisseHack** [comment]
Should I create an issue in that repo then to track? I'm happy to do any troubleshooting etc. that might help figure out where the bug might be.

**Comment by RikkiGibson** [comment]
Yes I think that's a good idea. Go ahead and link to this issue from that one.

**Comment by Youssef1313** [comment]
@RikkiGibson Should this issue be closed?
